osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs

```diff
@@ -19,39 +20,137 @@ namespace osu.Game.Rulesets.Taiko.Difficulty
 {
     public class TaikoDifficultyCalculator : DifficultyCalculator
     {
-        private const double star_scaling_factor = 0.04125;
+        private const double rhythm_skill_multiplier = 0.014;
+        private const double colour_skill_multiplier = 0.01;
+        private const double stamina_skill_multiplier = 0.02;
+
+        private readonly TaikoDifficultyHitObjectRhythm[] commonRhythms =
+        {
+            new TaikoDifficultyHitObjectRhythm(1, 1, 0.0, true),
+            new TaikoDifficultyHitObjectRhythm(2, 1, 0.3, false),
+            new TaikoDifficultyHitObjectRhythm(1, 2, 0.5, false),
+            new TaikoDifficultyHitObjectRhythm(3, 1, 0.3, false),
+            new TaikoDifficultyHitObjectRhythm(1, 3, 0.35, false),
+            new TaikoDifficultyHitObjectRhythm(3, 2, 0.6, false),
+            new TaikoDifficultyHitObjectRhythm(2, 3, 0.4, false),
+            new TaikoDifficultyHitObjectRhythm(5, 4, 0.5, false),
+            new TaikoDifficultyHitObjectRhythm(4, 5, 0.7, false)
+        };
 
         public TaikoDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
             : base(ruleset, beatmap)
         {
         }
 
+        private double simpleColourPenalty(double staminaDifficulty, double colorDifficulty)
+        {
+            if (colorDifficulty <= 0) return 0.79 - 0.25;
+
+            return 0.79 - Math.Atan(staminaDifficulty / colorDifficulty - 12) / Math.PI / 2;
+        }
+
+        private double norm(double p, double v1, double v2, double v3)
+        {
+            return Math.Pow(
+                Math.Pow(v1, p) +
+                Math.Pow(v2, p) +
+                Math.Pow(v3, p)
+                , 1 / p);
+        }
+
+        private double rescale(double sr)
+        {
+            if (sr < 0) return sr;
+
+            return 10.43 * Math.Log(sr / 8 + 1);
+        }
+
+        private double locallyCombinedDifficulty(double staminaPenalty, Skill colour, Skill rhythm, Skill stamina1, Skill stamina2)
```

Comment:
* would rename `stamina{1,2}` to `stamina{Left,Right}`
* kind of sketchy how all these could be properly-typed arguments here but are passed as the base `Skill` instead - potential for accidental misuse